_posts/2019-02-06-Age-Targeting-Part2.md

```diff
@@ -0,0 +1,255 @@
+---
+layout: post
+title: Increase your target audience via ML on Spark [part 2]
+subtitle: Preprocessing, train and predict
+category: data
+tags: [adtech, howto, spark, ml]
+author: Alessio Borgheresi
+header-img: "images/Age-Targeting/man_woman_question.jpg"
+---
+
+In mobile advertising it is possible to target specific audiences based on some characteristics of the users, such as the age. 
+However, this information is not available for every users.
+We build a supervised machine learning model to extend the knowledge over the whole dataset.
+In the previous [post](https://adello.github.io/Age-Targeting-Part1/) we described how the reliable ground truth dataset was created.
+It looks like this:
+<p align="center"> <img src="../images/Age-Targeting/table_ground_truth_age.png"> </p>
+For these devices we know the age and we convert it into 5 ranges: 13_17, 18_24, 25_34, 35_54, 55+.
+These are the classes used to train the machine learning model.
+
+To deal with the huge amount of mobile traffic we make use of spark, plus several hive queries to create the raw data. 
+The spark ML library require that the target column is an integer in the range [0, number of classes) and provide the StringIndexer class to transform your column. 
+
+```python
+from pyspark.ml.feature import StringIndexer
+
+label_stringidx = StringIndexer(inputCol=target_col, outputCol=label_col)
+df = label_stringidx.fit(df).transform(df)
+```
+
+**Deal with the unbalance of the classes**<br/>
+Now let's have a look at the distribution of the ground truth:
+<p align="center"> <img src="../images/Age-Targeting/table_age_frequency.png"> </p>
+As expected, some classes are less frequent then others.
+Usually the main problem with an unbalanced dataset is that during the training phase the model will not be able to learn the optimal patterns to distinguish the lower occurence class by the more frequent one.
```

Comment:
An unbalanced dataset can lead to problems in the training. The cannot learn the optimal patterns to distinguish...